js/src/main/scala/scalajson/ast/JValue.scala

```diff
@@ -48,86 +48,281 @@ final case class JString(value: String) extends JValue {
+   */
+  def toBigInt: Option[BigInt]
+
+  /**
+   * Return this number as a [[scala.Long]] if it's a valid [[scala.Long]].
+   */
+  def toLong: Option[Long]
 
-  def apply(value: Integer): JNumber = new JNumber(value.toString)
+  /**
+   * Return this number as a [[scala.Int]] if it's a valid [[scala.Int]].
+   */
+  def toInt: Option[Int]
 
-  def apply(value: Long): JNumber = new JNumber(value.toString)
+  /**
+   * Convert this number to its best [[scala.Double]] approximation.
+   */
+  def toDouble: Double
+
+  override def toJsAny: js.Any = toDouble match {
+    case n if n.isNaN => null
+    case n if n.isInfinity => null
+    case n => n
+  }
+}
 
-  def apply(value: BigInt): JNumber = new JNumber(value.toString())
+/**
+ * Constructors, type class instances, and other utilities for [[JNumber]].
+ */
+object JNumber {
+  def apply(value: BigDecimal): JNumber = JBigDecimal(value)
+  def apply(value: BigInt): JNumber = JBigInt(value)
+  def apply(value: Long): JNumber = JLong(value)
+  def apply(value: Int): JNumber = JInt(value)
+  def apply(value: Short): JNumber = JInt(value.toInt)
 
-  def apply(value: BigDecimal): JNumber = new JNumber(value.toString())
+  /**
+    * @param value
+    * @return Will return a [[JNull]] if value is a Nan or Infinity
+    */
+  def apply(value: Double): JValue = fromDouble(value)
 
   /**
     * @param value
-    * @return Will return a JNull if value is a Nan or Infinity
+    * @return Will return a [[JNull]] if value is a Nan or Infinity
     */
-  def apply(value: Double): JValue = value match {
+  def fromDouble(value: Double): JValue = value match {
     case n if n.isNaN => JNull
     case n if n.isInfinity => JNull
-    case _ => new JNumber(value.toString)
+    case _ => JDouble(value)
   }
 
   /**
     * @param value
-    * @return Will return a JNull if value is a Nan or Infinity
+    * @return Will return a [[JNull]] if value is a Nan or Infinity
     */
-  def apply(value: Float): JValue = value match {
+  def apply(value: Float): JValue = fromFloat(value)
+
+  /**
+    * @param value
+    * @return Will return a [[JNull]] if value is a Nan or Infinity
+    */
+  def fromFloat(value: Float): JValue = value match {
     case n if java.lang.Float.isNaN(n) => JNull
     case n if n.isInfinity => JNull
-    case _ => new JNumber(value.toString)
+    case _ => JFloat(value)
+  }
+
+  def apply(value: Array[Char]): Option[JNumber] = fromString(new String(value))
+  def apply(value: String): Option[JNumber] = fromString(value)
+
+  def fromString(value: String): Option[JNumber] = value match {
+    case jNumberRegex(_ *) => Option(JStringDecimal(value))
+    case _ => None
   }
 
-  def apply(value: String): Option[JNumber] =
-    fromString(value)
+  def unapply(value: JNumber): Option[String] = Option(value.stringValue)
+
+  private[ast] def jnumberEquals(value: JNumber, obj: Any): Boolean =
+    (value, obj) match {
```

Comment:
Seeing as `jnumberEquals` might be hot in certain kinds of applications: Could this be expressed in an allocation-free manner (possibly sacrificing a bit of conciseness)?